### PR TITLE
[TypeDeclaration] Remove StrictArrayParamDimFetchRector and StrictStringParamConcatRector

### LINKED ISSUE #8409: `StrictArrayParamDimFetchRector` assumes everything is an array while it can be `ArrayAccess`
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | 0.19.0      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.com/demo/a9b7d0bf-c5fa-4d77-b87a-a923c83c748a

```php
<?php

function test($data)
{
    $foo = $data['foo'];
    var_dump($foo);
}

test(new ArrayObject(['foo' => 'bar']));
```

### Responsible rules

* `StrictArrayParamDimFetchRector`

## Expected Behavior

In this case no `array` typehint should be set, because the actua passed param is an instance of `ArrayObject` (/implements `ArrayAccess`).
The original code works fine: https://3v4l.org/RbY8B
The result of Rector doesn't: https://3v4l.org/cWK1Z

Based on "Rector should only make changes of which it is confident" I would actually say that this rule "can't exist". Because a "dim fetch" operation can happen on more types / things than only an array so there is absolutely no guarantee that the variable would actually be an array / must be an array.

### PR DESCRIPTION
**Pull request by samsonasik**
While it usefull, it will buggy as it doesn't know the external caller, there is already rules that ensure verify the params usage, eg: `ParamTypeByMethodCallTypeRector` so I think these rules can be removed:

- StrictArrayParamDimFetchRector
- StrictStringParamConcatRector

Closes https://github.com/rectorphp/rector/issues/8409

### CODE REVIEW

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by staabm** [comment]
these rules worked pretty well for us in the past, as we don't have these ArrayAccess use-cases in our apps.
I would miss them :).

having several rules infering the same types based on different indications is a good thing IMO, as increasing type-coverage is a task which takes a lot of time otherwise and each code-base works differently. therefore having multiple rector rules for this cases is helpful IMO.

maybe we could make `StrictArrayParamDimFetchRector` configurable so one decide whether it should infer `array` or `array|ArrayAccess`?

**Comment by samsonasik** [comment]
`StrictArrayParamDimFetchRector` can buggy when next usage is `isset()` that will require check both variable or/and its `ArrayDimFetch`. I am not sure if configurable is better, as still, the caller is way better to verify the param values, otherwise, the method is mostly dead code.

on `StrictStringParamConcatRector`, it buggy as it can be `int, string, or stringable object, even null`

**Comment by staabm** [comment]
I see. 

maybe these rules can be combined in a "Risky" set, similar to how php-cs-fixer declares rules, which can have ugly side-effects but work most of the time

**Comment by samsonasik** [comment]
I will let @TomasVotruba decide 👍

**Comment by samsonasik** [comment]
This is one of possible issue with param rules  that not rely on external caller 

- https://getrector.com/demo/f8579121-6f81-4539-9a4d-c2df4905de06
- https://getrector.com/demo/830640fd-58fd-4b33-9a00-22e9f2cf0456

which can be `bool`, which should not happen.

**Comment by RobertMe** [comment]
I agree with @staabm, the rule is nice, if it works (which it does in most cases) :) But it is a "risky" rule. So as I already wrote in the original ticket, it doesn't really comply with Rectors "only make changes with confidence" / "only make changes which don't break the code".

In my case I was adding / executing the rules one by one and manually reviewing every change, so it's fine that these issues occur and I noticed them pretty quickly by review. But when running the whole set on a large codebase there is a massive chance that these issues will slip through (by not reviewing and/or not testing properly).

**Comment by RobertMe** [comment]
> This is one of possible issue with param rules that not rely on external caller
> 
> * https://getrector.com/demo/f8579121-6f81-4539-9a4d-c2df4905de06
> * https://getrector.com/demo/830640fd-58fd-4b33-9a00-22e9f2cf0456
> 
> which can be `bool`, which should not happen.

These are "nice" examples too. And I now remember I've seen the same issue with other rules as well. I for example had a case where a property could work both with an array as a (Doctrine) Collection. The constructor assigned a default value of `[]` and at least the `TypedPropertyFromStrictConstructorRector` rule only checked the constructor so just added the `array` typehint (based on the `[]`) assignment). While the class also had an `add`er method which did a `if ($this->prop instanceof Collection)` check which well, clearly showed the type wouldn't only be an `array`.

And as a sidenote: I've also been thinking about how Rector implements all these rules. And personally I think it's quite "messy" with all these rules which start out with the same thing (for example all return type rules collecting all returns in the function). And it might have been better if there was a single rule per thing (property, param, return) which collected all "usages" (so all usages of the property, all usages of the param, and all `return` statements) and would then use some "{property / param / return} type resolver" interface (/chain implementing an interface) to determine the type in every scenario, and only in case of all these "type resolvers" resolving to the same type per usage determine that this type is valid.

So with your example of https://getrector.com/demo/f8579121-6f81-4539-9a4d-c2df4905de06 you could have a multitude of "param type resolvers" for which one for example works based on strict comparison (I don't think Rector has a rule for that now?) and another which works based on array dim fetch. So the first "type resolver" based on strict comparison sees the `$param === false` statement and states the type can be `bool` / `false`, and for the other usage (`$param['test']`) it abstains from resolving a type. The second rule, based on array dim fetch, then abstains from resolving a type on the `$param === false` usage, but does resolve a type for `$param['test']`, being `array`.
So the ("top level") rule could then for example check whether every usage of the param does resolve to a type (so in this case one usage would result in `bool` and the in `array`, so all usages have a type resolved) and make a union out of it. While if a single usage doesn't resolve in a type it could skip this param completely for setting a type hint, as the "confidence" is lacking. And one could even make the rule configurable in whether unions might be created or not: i.e.: whether all usages must resolve to the same type in which case the return type would thus be a single type, or whether all usages might resolve to different types and making a union out of them.
And this chain of type resolvers could/should then be configurable so you can enable these one by one.

But I do obviously understand that this isn't how this is currently implemented in Rector and that changing this into something else would be a huge undertaking.

**Comment by samsonasik** [comment]
I understand what usefull the rules is, but the return type is different with param type, return type can be detected by value returned, but for param type, that is really rely on external call that call it, that's why there are rules:

- AddParamTypeFromPropertyTypeRector
- AddParamTypeBasedOnPHPUnitDataProviderRector
- ParamTypeByParentCallTypeRector
- ParamTypeByMethodCallTypeRector
- AddMethodCallBasedStrictParamTypeRector

that rely rely on external intervention that use/assigned from it.

**Comment by samsonasik** [comment]
Rebased 👍

**Comment by samsonasik** [comment]
Rebased 👍

File: src/Configuration/Levels/TypeCoverageLevel.php [hunk 1]
```diff
@@ -95,7 +93,5 @@ final class TypeCoverageLevel
         AddParamTypeFromPropertyTypeRector::class,
         MergeDateTimePropertyTypeDeclarationRector::class,
         PropertyTypeFromStrictSetterGetterRector::class,
-        StrictArrayParamDimFetchRector::class,
-        StrictStringParamConcatRector::class,
```
Comment (samsonasik) [inline src/Configuration/Levels/TypeCoverageLevel.php]: @TomasVotruba I can see it moved to "risky" level, I still prefer to just remove them, as while it usefull, param should cover from outside call/access (from method call or property fetch).

**Comment by samsonasik** [comment]
Rebased 👍

**Comment by samsonasik** [comment]
Closing as 1.0 released, and can't just be removed per https://github.com/rectorphp/rector-src/pull/5602#issuecomment-1937487301

**Pull request closed without merging**